### PR TITLE
INT-352 Port nexus3 script to chef

### PR DESCRIPTION
**Pull request by bigspotteddog**
https://issues.sonatype.org/browse/INT-352

Screencast: https://www.youtube.com/watch?v=99y0tePltFA

The Chef Nexus Repository Manager cookbook source is at:
https://github.com/sonatype/chef-nexus-repository-manager

Our motivation for using chef here is the Integrations team is creating reference deployments of Nexus Repository Manger and Nexus IQ Server on several platforms. Our hope is to use one cookbook for Nexus Repository Manager and one cookbook for Nexus IQ Server on all platforms.

We have worked the resulting image size down to 473MB; slightly larger than what it was before using chef, 466MB.

### CODE REVIEW

**Comment by patrickmslatteryvt** [comment]
So how do we know that there isn't a virus or other malware in that cookbook tar.gz you are sucking in from AWS?
I don't see any benefit in using Chef in this case, it's just adding another level of of indirection to the build process.
It also makes the image balloon up to 1.2GB in size!
```
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
sonatype/nexus3     chef                5d7c9958332b        5 seconds ago       1.22GB
sonatype/nexus3     latest              65e5c50fb48e        7 minutes ago       466MB
```
It's not my call but I wouldn't merge this.

**Comment by bigspotteddog** [comment]
@patrickmslatteryvt Thank you for the excellent feedback!

You are correct, downloading the cookbook.tar.gz from AWS is not secure, and it is temporary. I should have added detail to the PR description. Our nexus cookbook is private right now and we are downloading the cookbook.tar.gz from s3 until we make the nexus cookbook public.

Agree with your assessment that chef adds another level of indirection and that it makes the image larger. Our motivation for using chef here is the Integrations team is creating reference implementations of Nexus Repository Manger and Nexus IQ Server on several platforms, Amazon Web Services, Google Cloud Platform, Docker, Mesosphere, Red Hat Containers, etc. Our hope is to use this one cookbook on all of these platforms.

**Comment by patrickmslatteryvt** [comment]
So is this for a private implementation or for Sonatype themselves? I can't tell if you a Sonatype employee or contractor.
If for a private implementation then I think you would be better off forking the repo and making whatever changes you want there and build away to your hearts content :-)
If this is for an official Sonatype project then I'd say that having a single way to build all of the image types that you list is a great idea but it definitely needs to produce, at the very least, almost as good an image as would be produced by an optimized native build script for that image type.
The jump in final image size from 466MB to 1222MB would be a prime example of what not to do.
If the final image was say 500MB instead then I don't think very many people would care in the least how you built it.
It's been a long time since I last used Chef so I don't recall if you have to use a zipped cookbook or if you can simply point to a repo instead. The point being that if you provide (and consume) the cookbook as plain text files from a Github repo then no-one is going to wonder "What exactly is in that tar file?"

**Comment by bigspotteddog** [comment]
Closing this PR for now.

**Comment by whyjustin** [comment]
Hey @patrickmslatteryvt, I'm a Product Owner at Sonatype and would like to clarify some things. We are attempting to reduce costs in maintaining the runtime and application layer of multiple different Nexus Repository offerings. We chose to use Chef to provide an single instruction set for how Sonatype suggests to manage these two layers of a system.

I completely agree that transparency is necessary in generating a Docker image and is expected by the community. We will be open sourcing the Chef cookbook GitHub repository and providing a transparent approach to pulling the recipe into the Docker image to build out the internals. This will occur prior to any Pull Request being merged into master. Since Sonatype uses GitHub for communication and code review, some of this sausage making may bleed into a public forum before it is ready. This is what you are seeing.

Thanks for your diligence.

File: solo.json.erb [hunk 1]
```diff
@@ -0,0 +1,32 @@
+<%=
+require 'json'
+{
```
Comment (whyjustin) [inline solo.json.erb]: Can fail fast using `raise RuntimeError` when required variables are not supplied like in [0]

[0] https://github.com/sonatype/docker-nexus-iq-server/pull/1

File: solo.json.erb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  :run_list => [ 'recipe[nexus-repository-manager::docker]' ],
+  :java => {
+    :jdk_version => ENV['JAVA_VERSION'],
+    :java_home => ENV['JAVA_HOME'],
+    :install_flavor => 'oracle',
+    :oracle => {
+      :accept_oracle_download_terms => true
+    },
+    :jdk => {
+      '8' => {
```
Comment (whyjustin) [inline solo.json.erb]: How is this effected if a custom Java Version is supplied? This only handled v8.

File: Dockerfile [hunk 3]
```diff
@@ -22,17 +22,14 @@ LABEL vendor=Sonatype \
 
 ARG NEXUS_VERSION=3.5.2-01
```
Comment (whyjustin) [inline Dockerfile]: Let's standardize variable case between this and https://github.com/sonatype/docker-nexus-iq-server

File: solo.json.erb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        :x86_64 => {
+          :url => ENV['JAVA_URL'],
+          :checksum => ENV['JAVA_DOWNLOAD_HASH']
+        }.compact
+      }
+    }
+  },
+  'nexus-repository-manager': {
```
Comment (whyjustin) [inline solo.json.erb]: `anager': {` or `anager' => {`

File: Dockerfile (hunk 3 above)
Comment (bigspotteddog) [reply Dockerfile]: Kept this case to maintain compatibility with previous version of the Docker file.

File: solo.json.erb (hunk 1 above)
Comment (bigspotteddog) [reply solo.json.erb]: Added error checks for the three "required" variables.

File: solo.json.erb (hunk 2 above)
Comment (bigspotteddog) [reply solo.json.erb]: The java recipe cannot automatically download java 6 or 7 from Oracle. I was able to host the binary on my own server and have it install correctly; however, it appears that Java 1.8 is required.

If Java 1.8 is indeed required, this should probably be removed as an option.

File: solo.json.erb (hunk 4 above)
Comment (bigspotteddog) [reply solo.json.erb]: Changed to =>

File: solo.json.erb (hunk 2 above)
Comment (bigspotteddog) [reply solo.json.erb]: Keeping this as an option so the Java url and hash can be customized for Java 8.

**Comment by kellyrob99** [comment]
The jump in size is concerning and I'm not sure why it's happening at all. When the image is done and cleaned up, the size difference should be negligible...

**Comment by bigspotteddog** [comment]
@kellyrob99 We have worked the size down to 473MB vs 466MB before chef is added. I should have added a comment for this.

**Comment by patrickmslatteryvt** [comment]
I just built the latest image and it compares very favorably in size to the old image, it's only about 1% bigger.
```
$ docker images
REPOSITORY              TAG                 IMAGE ID            CREATED             SIZE
sonatype/nexus3         chef                4424065b7ffb        31 seconds ago      473MB
sonatype/nexus3         latest              65e5c50fb48e        10 days ago         466MB
```

File: Dockerfile [hunk 5]
```diff
@@ -12,51 +12,50 @@
...
-  JAVA_VERSION_MAJOR=8 \
-  JAVA_VERSION_MINOR=144 \
-  JAVA_VERSION_BUILD=01 \
-  JAVA_DOWNLOAD_HASH=090f390dda5b47b9b721c7dfaa008135
+    JAVA_VERSION_MAJOR=8 \
+    JAVA_VERSION_MINOR=144 \
+    JAVA_VERSION_BUILD=01 \
+    JAVA_DOWNLOAD_HASH=090f390dda5b47b9b721c7dfaa008135
+
+ENV JAVA_URL=http://download.oracle.com/otn-pub/java/jdk/${JAVA_VERSION_MAJOR}u${JAVA_VERSION_MINOR}-b${JAVA_VERSION_BUILD}/${JAVA_DOWNLOAD_HASH}/server-jre-${JAVA_VERSION_MAJOR}u${JAVA_VERSION_MINOR}-linux-x64.tar.gz \
```
Comment (whyjustin) [inline Dockerfile]: ENV can be declared above, no need for a new layer just for these declarations.

File: Dockerfile [hunk 6]
```diff
@@ -12,51 +12,50 @@
...
+    JAVA_DOWNLOAD_SHA256_HASH=e8a341ce566f32c3d06f6d0f0eeea9a0f434f538d22af949ae58bc86f2eeaae4
 
 # configure nexus runtime
 ENV SONATYPE_DIR=/opt/sonatype
 ENV NEXUS_HOME=${SONATYPE_DIR}/nexus \
```
Comment (whyjustin) [inline Dockerfile]: ENV can be declared above, no need for a new layer just for these declarations.

File: Dockerfile [hunk 7]
```diff
@@ -12,51 +12,50 @@
...
+       --json-attributes /var/chef/solo.json \
+    && rpm -qa *chef* | xargs rpm -e \
+    && rpm --rebuilddb \
+    && rm -rf /etc/chef \
+    && rm -rf /opt/chefdk \
+    && rm -rf /var/cache/yum \
+    && rm -rf /var/chef
 
 # configure nexus
 RUN sed \
```
Comment (whyjustin) [inline Dockerfile]: Why is this still needed? Shouldn't it be managed with Chef Solo?

File: solo.json.erb [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  :run_list => [ 'recipe[nexus_repository_manager::docker]' ],
+  :java => {
+    :jdk_version => ENV['JAVA_VERSION_MAJOR'],
+    :java_home => ENV['JAVA_HOME'],
+    :install_flavor => 'oracle',
+    :oracle => {
+      :accept_oracle_download_terms => true
+    },
+    :jdk => {
+      '8' => {
```
Comment (whyjustin) [inline solo.json.erb]: `:8` rather than `'8'`

File: spec/Dockerfile_spec.rb [hunk 9]
```diff
@@ -0,0 +1,43 @@
+# Copyright (c) 2017-present Sonatype, Inc.
```
Comment (whyjustin) [inline spec/Dockerfile_spec.rb]: Generally the copyright date is consistent across the project, so this can be 2016.

File: Dockerfile (hunk 5 above)
Comment (patrickmslatteryvt) [reply Dockerfile]: Actually you can't declare and consume an ENV in the same layer.

Comment (whyjustin) [reply Dockerfile]: 👍  @patrickmslatteryvt, thanks

File: solo.json.erb (hunk 8 above)
Comment (fernau) [reply solo.json.erb]: i don't think ruby allows `:8`

File: spec/Dockerfile_spec.rb (hunk 9 above)
Comment (bigspotteddog) [reply spec/Dockerfile_spec.rb]: Changed to 2016 (e4188af).

File: Dockerfile (hunk 6 above)
Comment (bigspotteddog) [reply Dockerfile]: SONATYPE_DIR is consumed in NEXUS_HOME. I thought the same thing, but it did not work.

File: solo.json.erb (hunk 8 above)
Comment (bigspotteddog) [reply solo.json.erb]: Seems to work as 8, not as :8. Changed to :'8'

File: Dockerfile (hunk 7 above)
Comment (bigspotteddog) [reply Dockerfile]: Removed both sed calls and moved this logic into chef (8acfc2d, dc3ec34).

File: Dockerfile [hunk 10]
```diff
@@ -12,73 +12,56 @@
...
+    && rm -rf /var/cache/yum \
+    && rm -rf /var/chef
 
 VOLUME ${NEXUS_DATA}
 
 EXPOSE 8081
 USER nexus
-WORKDIR ${NEXUS_HOME}
 
 ENV INSTALL4J_ADD_VM_PARAMS="-Xms1200m -Xmx1200m -XX:MaxDirectMemorySize=2g -Djava.util.prefs.userRoot=${NEXUS_DATA}/javaprefs"
```
Comment (bigspotteddog) [inline Dockerfile]: Attempted to move this line into chef. Apparently, chef does not set system environment variables that are outside of it process or subprocesses. Tried several workarounds (magic_shell, FieEdit on /etc/environment, others). The $INSTALL4J_ADD_VM_PARAMS was not available.

File: Dockerfile [hunk 11]
```diff
@@ -48,7 +48,7 @@ ADD solo.json.erb /var/chef/solo.json.erb
 RUN curl -L https://www.getchef.com/chef/install.sh | bash \
     && /opt/chef/embedded/bin/erb /var/chef/solo.json.erb > /var/chef/solo.json \
     && chef-solo \
-       --recipe-url https://github.com/sonatype/chef-nexus-repository-manager/releases/download/${NEXUS_REPOSITORY_MANAGER_COOKBOOK_VERSION}/chef-nexus-repository-manager.tar.gz \
+       --recipe-url https://s3.amazonaws.com/jcava-int-public/nexus-repository-manager-cookbook.tar.gz \
```
Comment (fernau) [inline Dockerfile]: i think this is test code

Comment (bigspotteddog) [reply Dockerfile]: Changed back to correct url (ec59126).

File: Dockerfile [hunk 12]
```diff
@@ -12,73 +12,56 @@
...
 
-RUN useradd -r -u 200 -m -c "nexus role account" -d ${NEXUS_DATA} -s /bin/false nexus \
-  && mkdir -p ${NEXUS_DATA}/etc ${NEXUS_DATA}/log ${NEXUS_DATA}/tmp ${SONATYPE_WORK} \
-  && ln -s ${NEXUS_DATA} ${SONATYPE_WORK}/nexus3 \
-  && chown -R nexus:nexus ${NEXUS_DATA}
+# Install using chef-solo
+RUN curl -L https://www.getchef.com/chef/install.sh | bash \
+    && /opt/chef/embedded/bin/erb /var/chef/solo.json.erb > /var/chef/solo.json \
+    && chef-solo \
+       --recipe-url https://github.com/sonatype/chef-nexus-repository-manager/releases/download/${NEXUS_REPOSITORY_MANAGER_COOKBOOK_VERSION}/chef-nexus-repository-manager.tar.gz \
```
Comment (bradbeck) [inline Dockerfile]: I would suggest exposing the recipe url as a build argument for cases where we want to run with a custom recipe. Similar to what is currently being don with NEXUS_DOWNLOAD_URL.

Comment (bigspotteddog) [reply Dockerfile]: Exposed the recipe version and url as build args. Thanks!

**Pull request merged**